File: core/api/golang/lib/bulk_command_execution/bulk_command_serializer.go
/*
 *    Copyright 2021 Kurtosis Technologies Inc.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 *
 */

package bulk_command_execution

import (
	"encoding/json"
	"github.com/kurtosis-tech/kurtosis-client/golang/lib/bulk_command_execution/v0_bulk_command_api"
	"github.com/kurtosis-tech/stacktrace"
)

const (
	latestSchemaVersion = V0
)

type VersionedBulkCommandsDocument struct {
	SchemaVersion SchemaVersion `json:"schemaVersion"`
}

type serializableBulkCommandsDocument struct {
	VersionedBulkCommandsDocument
	Body	interface{}			`json:"body"`
}

type BulkCommandSerializer struct {}

func NewBulkCommandSerializer() *BulkCommandSerializer {
	return &BulkCommandSerializer{}
}

func (serializer BulkCommandSerializer) Serialize(bulkCommands v0_bulk_command_api.V0BulkCommands) ([]byte, error) {
	toSerialize := serializableBulkCommandsDocument{
		VersionedBulkCommandsDocument: VersionedBulkCommandsDocument{
			SchemaVersion: latestSchemaVersion,
		},
		Body:                          bulkCommands,
	}
	bytes, err := json.Marshal(toSerialize)
	if err != nil {
		return nil, stacktrace.Propagate(err, "An error occurred serializing bulk commands to bytes")
	}
	return bytes, nil
}

File: core/api/typescript/src/lib/bulk_command_execution/bulk_command_serializer.ts
import { V0BulkCommands } from "./v0_bulk_command_api/v0_bulk_commands";
import { SchemaVersion } from "./bulk_command_schema_version";
import { ok, err, Result } from "neverthrow";

const LATEST_SCHEMA_VERSION: SchemaVersion = SchemaVersion.V0;

class VersionedBulkCommandsDocument {
    private readonly schemaVersion: SchemaVersion;
    
    constructor(schemaVersion: SchemaVersion) {
        this.schemaVersion = schemaVersion;
    }
}

class SerializableBulkCommandsDocument extends VersionedBulkCommandsDocument {
    private readonly body: any;
    
    constructor(schemaVersion: SchemaVersion, body: V0BulkCommands) {
        super(schemaVersion);
        this.body = body;
    }
}


class BulkCommandSerializer {

    constructor (){}

    public serialize(bulkCommands: V0BulkCommands): Result<Uint8Array | string, Error> {
        const toSerialize: SerializableBulkCommandsDocument = new SerializableBulkCommandsDocument(LATEST_SCHEMA_VERSION, bulkCommands);
        
        let bytes: string;
        try {
            bytes = JSON.stringify(toSerialize);
        } catch (jsonErr) {
            if(jsonErr instanceof Error){
                return err(jsonErr);
            }
            return err(new Error("Stringify-ing SerializableBulkCommandsDocument object threw an exception, but " +
                "it's not an Error so we can't report any more information than this"));
        }

        return ok(bytes);
    }
}
